bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/RuleEngine.java

```diff
@@ -0,0 +1,426 @@
+        Collection<Rule> result = new ArrayList<Rule>(10);
+        for (Iterator<RuleImpl> it = rules.values().iterator(); it.hasNext();) {
+            RuleImpl r = it.next();
+            if (tag != null) {
+                Set<String> tags = r.getTags();
+                if (tags != null && tags.contains(tag)) {
+                    result.add(new RuleImpl(r));
+                }
+            } else {
+                result.add(new RuleImpl(r));
+            }
+        }
+        return result;
+    }
+
+    public Collection<Rule> getRules(String filter) {
+        Collection<Rule> result = new ArrayList<Rule>(10);
+        for (Iterator<RuleImpl> it = rules.values().iterator(); it.hasNext();) {
+            RuleImpl r = it.next();
+            result.add(new RuleImpl(r));
+        }
+        return result;
+    }
+
+    protected Collection<String> getScopeIds() {
+        Set<String> result = new HashSet<String>(10);
+        for (Iterator<RuleImpl> it = rules.values().iterator(); it.hasNext();) {
+            RuleImpl r = it.next();
+            String id = r.getScopeIdentifier();
+            if (id != null) {
+                result.add(id);
+            }
+        }
+        return result;
+    }
+
+    /**
+     *
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#addingService(org.osgi.framework.ServiceReference)
+     */
+    @Override
+    public ModuleHandlerFactory addingService(ServiceReference/* <ModuleHandlerFactory> */reference) {
+        ModuleHandlerFactory msf = (ModuleHandlerFactory) bc.getService(reference);
+        Collection<String> moduleTypes = msf.getTypes();
+        for (Iterator<String> it = moduleTypes.iterator(); it.hasNext();) {
+            String moduleTypeName = it.next();
+            if (moduleHandlerFactories.get(moduleTypeName) != null) {
+                // log.error
+            } else {
+                moduleHandlerFactories.put(moduleTypeName, msf);
+            }
+
+        }
+        return msf;
+    }
+
+    /**
+     *
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#modifiedService(org.osgi.framework.ServiceReference,
+     *      java.lang.Object)
+     */
+    @Override
+    public void modifiedService(ServiceReference/* <ModuleHandlerFactory> */reference, /* ModuleHandlerFactory */
+            Object service) {
+        // TODO Auto-generated method stub
+
+    }
+
+    /**
+     * @see org.osgi.util.tracker.ServiceTrackerCustomizer#removedService(org.osgi.framework.ServiceReference,
+     *      java.lang.Object)
+     */
+    @Override
+    public void removedService(ServiceReference/* <ModuleHandlerFactory> */reference, /* ModuleHandlerFactory */
+            Object service) {
+        Collection<String> moduleTypes = ((ModuleHandlerFactory) service).getTypes();
+        for (Iterator<String> it = moduleTypes.iterator(); it.hasNext();) {
+            String moduleTypeName = it.next();
+            moduleHandlerFactories.remove(moduleTypeName);
```

Comment:
we have to remove the handlers from the modules and deactivate the respective rules.